diff -Nur suck-4.2.4.old/sample/get.news.generic suck-4.2.4/sample/get.news.generic
--- suck-4.2.4.old/sample/get.news.generic	Fri Feb 16 21:38:21 2001
+++ suck-4.2.4/sample/get.news.generic	Fri Feb 16 21:37:54 2001
@@ -18,14 +18,19 @@
 MSGDIR=${BASEDIR}/Msgs			# where to put MultiFile articles when getting them
 
 OUTGOING=${SPOOLDIR}/out.going/${REMOTE_HOST}	# location of the list of articles to upload
-SCRIPT=${BASEDIR}/put.news		# my filter for rpost
 OUTFILE=/tmp/tmp$$			# used by rpost as article after it is filtered
 LOCKFILE=${BASEDIR}/getnews.lock	# lock file to prevent multiple instances of script
 
 TESTHOST=testhost
-RPOST=rpost
 SUCK=suck
-TESTHOST=testhost
+
+# for INN 2.x with Storage API use:
+SCRIPT=${BASEDIR}/put.news.sm		# my filter for rpost
+RPOST=rpost				# rpost command
+# for INN <2.3 without storage API use:
+#SCRIPT=${BASEDIR}/put.news		# my filter for rpost
+#RPOST="rpost -p ${SPOOLDIR}"		# rpost command
+
 
 # if we are already running, abort
 
@@ -71,7 +76,7 @@
 		if [ $? -ne 0 ]; then
 			echo "Remote posting host not responding"
 		else
-			${RPOST} ${REMOTE_HOST} -d -b ${OUTGOING} -p ${SPOOLDIR} -f \$\$o=${OUTFILE} ${SCRIPT} \$\$i ${OUTFILE}
+			${RPOST} ${REMOTE_HOST} -d -b ${OUTGOING} -f \$\$o=${OUTFILE} ${SCRIPT} \$\$i ${OUTFILE}
 
 			if [ $? -ne 0 ]; then
 				echo "Error remote posting"
diff -Nur suck-4.2.4.old/sample/get.news.inn suck-4.2.4/sample/get.news.inn
--- suck-4.2.4.old/sample/get.news.inn	Fri Feb 16 21:38:21 2001
+++ suck-4.2.4/sample/get.news.inn	Fri Feb 16 21:36:17 2001
@@ -22,13 +22,18 @@
 OUTGOING=${SPOOLDIR}/out.going/${SITE}	# location of the list of articles to upload
 OUTGOINGNEW=${OUTGOING}.new		# file to contain the list temporarily
 OUTGOINGFAIL=${OUTGOINGNEW}.fail	# file with failed xfers
-SCRIPT=${BASEDIR}/put.news		# my filter for rpost
 OUTFILE=/tmp/tmp$$			# used by rpost as article after it is filtered
 LOCKFILE=${BASEDIR}/getnews.lock	# lock file to prevent multiple instances of script
 NEWSGROUP=news				# which group owns the file in out.going, typically either news or uucp.
 
+# for INN 2.x with Storage API use:
+SCRIPT=${BASEDIR}/put.news.sm		# my filter for rpost
+RPOST=rpost				# rpost command
+# for INN <2.3 without storage API use:
+#SCRIPT=${BASEDIR}/put.news		# my filter for rpost
+#RPOST="rpost -p ${SPOOLDIR}"		# rpost command
+
 TESTHOST=testhost
-RPOST=rpost
 SUCK=suck
 
 # if we are already running, abort 
@@ -95,7 +100,7 @@
 				fi
 
 				# outgoing messages to post
-				${RPOST} ${REMOTE_HOST} -d -b ${OUTGOINGNEW} -p ${SPOOLDIR} -f \$\$o=${OUTFILE} ${SCRIPT} \$\$i ${OUTFILE}
+				${RPOST} ${REMOTE_HOST} -d -b ${OUTGOINGNEW} -f \$\$o=${OUTFILE} ${SCRIPT} \$\$i ${OUTFILE}
 
 				ERRLEV=$?
 				if [ ${ERRLEV} -eq 0 ]; then
